File: android/app/src/test/java/com/fitlog/app/InsightsDomainTest.kt
package com.fitlog.app

import com.fitlog.app.domain.Comparisons
import com.fitlog.app.domain.Insights
import com.fitlog.app.domain.Progress
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class InsightsDomainTest {

    private val vectors = TestVectors.load("insights.json").jsonObject

    private fun setInput(element: kotlinx.serialization.json.JsonElement): Comparisons.SetInput {
        val item = element.jsonObject
        return Comparisons.SetInput(
            exerciseId = item.getValue("exercise_id").jsonPrimitive.content,
            sessionId = item.getValue("session_id").jsonPrimitive.content,
            startedAtMs = item.getValue("started_at_ms").jsonPrimitive.content.toLong(),
            weightKg = item.getValue("weight_kg").jsonPrimitive.doubleOrNull,
            reps = item.getValue("reps").jsonPrimitive.intOrNull,
            isWarmup = item.getValue("is_warmup").jsonPrimitive.intOrNull == 1,
        )
    }

    @Test
    fun `los consejos coinciden con los vectores compartidos`() {
        vectors.getValue("cases").jsonArray.forEach { element ->
            val testCase = element.jsonObject
            val name = testCase.getValue("name").jsonPrimitive.content
            val groups = testCase.getValue("groups").jsonObject.mapValues { it.value.jsonPrimitive.content }
            val rangeObject = testCase.getValue("range").jsonObject

            val tips = Insights.build(
                Insights.Input(
                    sets = testCase.getValue("sets").jsonArray.map { setInput(it) },
                    groupByExercise = groups,
                    range = Progress.Range(
                        fromMs = rangeObject.getValue("from_ms").jsonPrimitive.doubleOrNull?.toLong(),
                        toMs = rangeObject.getValue("to_ms").jsonPrimitive.content.toLong(),
                    ),
                    periodDays = testCase.getValue("period_days").jsonPrimitive.intOrNull ?: 30,
                )
            )

            val expected = testCase.getValue("expected").jsonArray
            assertEquals("$name: cantidad de consejos", expected.size, tips.size)

            expected.forEachIndexed { index, expectedElement ->
                val item = expectedElement.jsonObject
                val tip = tips[index]

                assertEquals("$name: kind", item.getValue("kind").jsonPrimitive.content, tip.kind.name.lowercase())
                assertEquals(
                    "$name: severity",
                    item.getValue("severity").jsonPrimitive.content,
                    tip.severity.name.lowercase(),
                )
                assertEquals(
                    "$name: subject",
                    if (item.getValue("subject") is JsonNull) {
                        null
                    } else {
                        item.getValue("subject").jsonPrimitive.content
                    },
                    tip.subject,
                )
                assertEquals(
                    "$name: value",
                    item.getValue("value").jsonPrimitive.doubleOrNull ?: 0.0,
                    tip.value,
                    1e-9,
                )
                assertEquals("$name: message", item.getValue("message").jsonPrimitive.content, tip.message)
            }
        }
    }

    @Test
    fun `el formato de numeros coincide con la regla compartida`() {
        assertEquals("25", Insights.formatNumber(25.0))
        assertEquals("57.1", Insights.formatNumber(57.14))
        assertEquals("33.3", Insights.formatNumber(33.333))
    }

    @Test
    fun `los umbrales documentados estan fijados`() {
        assertEquals(3, Insights.PROGRESS_MIN_SESSIONS)
        assertEquals(0.5, Insights.STAGNATION_MAX_VARIATION_PCT, 1e-9)
        assertEquals(50.0, Insights.IMBALANCE_MIN_SHARE_PCT, 1e-9)
        assertEquals(3.0, Insights.CONSISTENCY_HIGH_PER_WEEK, 1e-9)
        assertEquals(1.5, Insights.CONSISTENCY_LOW_PER_WEEK, 1e-9)
        assertEquals(30.0, Insights.INCOMPLETE_MIN_SHARE_PCT, 1e-9)
        assertTrue(Insights.TIP_LIMIT > 0)
    }
}

File: web/src/test/insightsDomain.test.ts
import { describe, expect, it } from 'vitest';

import type { ComparisonSetInput } from '@/domain/comparisons';
import { buildTips, formatNumber, THRESHOLDS, type InsightsInput } from '@/domain/insights';

import vectors from '@shared/test-vectors/insights.json';

interface VectorSet {
  exercise_id: string;
  session_id: string;
  started_at_ms: number;
  weight_kg: number | null;
  reps: number | null;
  is_warmup: number;
}

interface VectorCase {
  name: string;
  sets: VectorSet[];
  groups: Record<string, string>;
  range: { from_ms: number | null; to_ms: number };
  period_days: number;
  expected: {
    kind: string;
    severity: string;
    subject: string | null;
    value: number;
    message: string;
  }[];
}

function toInput(testCase: VectorCase): InsightsInput {
  const sets: ComparisonSetInput[] = testCase.sets.map((set) => ({
    exerciseId: set.exercise_id,
    sessionId: set.session_id,
    startedAtMs: set.started_at_ms,
    weightKg: set.weight_kg,
    reps: set.reps,
    isWarmup: set.is_warmup === 1,
  }));

  return {
    sets,
    groupByExercise: testCase.groups,
    range: { fromMs: testCase.range.from_ms, toMs: testCase.range.to_ms },
    periodDays: testCase.period_days,
  };
}

describe('buildTips', () => {
  it.each(vectors.cases as VectorCase[])('$name', (testCase) => {
    const tips = buildTips(toInput(testCase));

    expect(tips).toHaveLength(testCase.expected.length);
    tips.forEach((tip, index) => {
      const expected = testCase.expected[index];
      expect(tip.kind).toBe(expected?.kind);
      expect(tip.severity).toBe(expected?.severity);
      expect(tip.subject).toBe(expected?.subject);
      expect(tip.value).toBeCloseTo(expected?.value ?? 0, 9);
      expect(tip.message).toBe(expected?.message);
    });
  });
});

describe('formatNumber', () => {
  it('imprime enteros sin decimales y el resto con un decimal', () => {
    expect(formatNumber(25)).toBe('25');
    expect(formatNumber(57.14)).toBe('57.1');
    expect(formatNumber(33.333)).toBe('33.3');
  });
});

describe('umbrales', () => {
  it('expone los umbrales documentados', () => {
    expect(THRESHOLDS.progressMinSessions).toBe(3);
    expect(THRESHOLDS.stagnationMaxVariationPct).toBe(0.5);
    expect(THRESHOLDS.imbalanceMinSharePct).toBe(50);
    expect(THRESHOLDS.consistencyHighPerWeek).toBe(3);
    expect(THRESHOLDS.consistencyLowPerWeek).toBe(1.5);
    expect(THRESHOLDS.incompleteMinSharePct).toBe(30);
  });
});
